kernel/src/main/java/org/kframework/compile/ResolveSemanticCasts.java

```diff
@@ -116,24 +124,67 @@ public void apply(KApply v) {
 
     K transform(K term) {
         return new TransformK() {
+            private Sort sort;
+
             @Override
-            public K apply(KApply k) {
+            public K apply(K k) {
+                final Sort oldSort = sort;
+                K applied;
                 if (casts.contains(k)) {
-                    return super.apply(k.klist().items().get(0));
+                    KApply kapp = (KApply)k;
+                    sort = Outer.parseSort(getSortNameOfCast(kapp));
+                    applied = apply(kapp.items().get(0));
+                } else {
+                    sort = null;
+                    applied = super.apply(k);
                 }
-                return super.apply(k);
-            }
-
-            @Override
-            public K apply(KVariable k) {
-                if (varToTypedVar.containsKey(k)) {
-                    return varToTypedVar.get(k);
+                if (oldSort != null) {
+                  return new AddAtt(a -> a.add(Sort.class, oldSort)).apply(applied);
                 }
-                return super.apply(k);
+                if (varToTypedVar.containsKey(applied)) {
+                  return varToTypedVar.get(applied);
+                }
+                return applied;
             }
         }.apply(term);
     }
 
+    public static class AddAtt extends TransformK {
+        private final UnaryOperator<Att> f;
+
+        public AddAtt(UnaryOperator<Att> f) {
+            this.f = f;
+        }
+
+        @Override
+        public K apply(KApply kapp) {
+            return KApply(kapp.klabel(), kapp.klist(), f.apply(kapp.att()));
+        }
+        @Override
+        public K apply(KRewrite rew) {
+            return KRewrite(rew.left(), rew.right(), f.apply(rew.att()));
+        }
+        @Override
+        public K apply(KToken tok) {
+            return KToken(tok.s(), tok.sort(), f.apply(tok.att()));
+        }
```

Comment:
This code is probably more broadly useful, consider moving it to where TransformK is defined?